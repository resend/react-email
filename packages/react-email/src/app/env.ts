/* eslint-disable @typescript-eslint/no-non-null-assertion */
/** ONLY ACCESSIBLE ON THE SERVER */
export const emailsDirRelativePath = process.env.EMAILS_DIR_RELATIVE_PATH!;

/** ONLY ACCESSIBLE ON THE SERVER */
export const userProjectLocation = process.env.USER_PROJECT_LOCATION!;

/** ONLY ACCESSIBLE ON THE SERVER */
export const emailsDirectoryAbsolutePath =
  process.env.EMAILS_DIR_ABSOLUTE_PATH!;

export const isBuilding = process.env.NEXT_PUBLIC_IS_BUILDING === 'true';

export const isPreviewDevelopment =
  process.env.NEXT_PUBLIC_IS_PREVIEW_DEVELOPMENT === 'true';
