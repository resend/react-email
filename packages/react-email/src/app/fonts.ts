import { Inter } from 'next/font/google';
import Local from 'next/font/local';

export const inter = Inter({
  subsets: ['latin'],
  variable: '--font-inter',
  display: 'swap',
});

export const sfMono = Local({
  src: [
    {
      path: './fonts/SFMono/SFMonoLight.otf',
      weight: '300',
    },
    {
      path: './fonts/SFMono/SFMonoRegular.otf',
      weight: '400',
    },
    {
      path: './fonts/SFMono/SFMonoMedium.otf',
      weight: '500',
    },
    {
      path: './fonts/SFMono/SFMonoSemibold.otf',
      weight: '600',
    },
    {
      path: './fonts/SFMono/SFMonoBold.otf',
      weight: '700',
    },
    {
      path: './fonts/SFMono/SFMonoHeavy.otf',
      weight: '800',
    },
  ],
  variable: '--font-sf-mono',
  display: 'swap',
});
