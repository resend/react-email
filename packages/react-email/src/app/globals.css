@tailwind base;
@tailwind components;
@tailwind utilities;

.popup-open iframe {
  pointer-events: none;
}

nav > div > div > .line {
  display: none;
}
