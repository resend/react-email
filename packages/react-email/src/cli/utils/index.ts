export * from './preview';
export * from './tree';
