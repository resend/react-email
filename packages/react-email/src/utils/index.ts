export * from './types/as';
export * from './cn';
export * from './copy-text-to-clipboard';
export * from './language-map';
export * from './unreachable';
export * from './sanitize';
