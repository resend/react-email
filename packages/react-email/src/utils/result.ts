type Ok<T, _E> = {
  value: T;
};
type Error<_T, E> = {
  error: E;
};

/**
 * Do not destructure this object, it is meant to have all fields together
 * in the same object
 */
export type Result<T, E> = Ok<T, E> | Error<T, E>;

export function isErr<T, E>(result: Result<T, E>): result is Error<T, E> {
  return 'error' in result;
}

export function isOk<T, E>(result: Result<T, E>): result is Ok<T, E> {
  return 'value' in result && !('error' in result);
}

export function mapResult<T, E, B>(
  result: Result<T, E>,
  callback: (value: T) => B,
): Result<B, E> {
  if (isOk(result)) {
    return ok(callback(result.value));
  }

  return result;
}

export function ok<T, E>(value: NoInfer<T>): Ok<T, E>;
// biome-ignore lint/suspicious/noConfusingVoidType: This is required for void return types on functions that can still error
export function ok<T extends void = void, E = never>(value: void): Ok<void, E>;
export function ok<T, E>(value: NoInfer<T>): Ok<T, E> {
  return {
    value,
  };
}

export function err<T, E>(error: NoInfer<E>): Error<T, E>;
// biome-ignore lint/suspicious/noConfusingVoidType: This is required for void return types on functions that can still error
export function err<T, E extends void = void>(error: void): Error<T, void>;
export function err<T, E>(error: NoInfer<E>): Error<T, E> {
  return {
    error,
  };
}
