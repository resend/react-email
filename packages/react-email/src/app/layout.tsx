import type { Metadata } from 'next';
import './globals.css';
import { EmailsProvider } from '../contexts/emails';
import { emailsDirectoryAbsolutePath } from '../utils/emails-directory-absolute-path';
import { getEmailsDirectoryMetadata } from '../utils/get-emails-directory-metadata';
import { inter } from './inter';

export const metadata: Metadata = {
  title: 'React Email',
};

export const dynamic = 'force-dynamic';

const RootLayout = async ({ children }: { children: React.ReactNode }) => {
  const emailsDirectoryMetadata = await getEmailsDirectoryMetadata(
    emailsDirectoryAbsolutePath,
  );

  if (typeof emailsDirectoryMetadata === 'undefined') {
    throw new Error(
      `Could not find the emails directory specified under ${emailsDirectoryAbsolutePath}!`,
    );
  }

  return (
    <html className={inter.className} lang="en">
      <body className="relative flex h-screen flex-col overflow-x-hidden bg-black leading-loose text-slate-11 selection:bg-cyan-5 selection:text-cyan-12">
        <EmailsProvider
          initialEmailsDirectoryMetadata={emailsDirectoryMetadata}
        >
          {children}
        </EmailsProvider>
        <div
          aria-hidden
          className="pointer-events-none absolute inset-0 bg-gradient-to-t from-slate-3"
        />
      </body>
    </html>
  );
};

export default RootLayout;
