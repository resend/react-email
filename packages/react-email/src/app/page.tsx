import path from 'node:path';
import Image from 'next/image';
import Link from 'next/link';
import { Button, Heading, Text } from '../components';
import { Shell } from '../components/shell';
import { emailsDirectoryAbsolutePath } from '../utils/emails-directory-absolute-path';
import logo from './logo.png';
import { getEmailsDirectoryMetadata } from '../actions/get-emails-directory-metadata';
import { getWarningsForEmails } from '../actions/get-warnings-for-emails';

const Home = async () => {
  const baseEmailsDirectoryName = path.basename(emailsDirectoryAbsolutePath);
  const emailsDirMetadata = await getEmailsDirectoryMetadata(
    emailsDirectoryAbsolutePath,
  );
  if (typeof emailsDirMetadata === 'undefined') {
    throw new Error(
      `Could not find the emails directory specified under ${emailsDirectoryAbsolutePath}!

This is most likely not an issue with the preview server. Maybe there was a typo on the "--dir" flag?`,
    );
  }

  const emailValidationWarnings = await getWarningsForEmails(emailsDirMetadata);

  return (
    <Shell emailValidationWarnings={emailValidationWarnings}>
      <div className="relative max-w-lg mx-auto p-8 flex items-center justify-center h-[inherit]">
        <div className="relative z-10 flex flex-col text-center items-center">
          <Image
            alt="React Email Icon"
            className="mb-8"
            height={144}
            src={logo}
            style={{
              borderRadius: 34,
              boxShadow: '0px 10px 200px 20px #2B7CA080',
            }}
            width={141}
          />
          <Heading as="h2" size="6" weight="medium">
            Welcome to React Email
          </Heading>
          <Text as="p" className="mt-2 mb-4">
            To start developing your emails, you can create a<br />
            <code className="text-slate-12">.jsx</code> or{' '}
            <code className="text-slate-12">.tsx</code> file under your{' '}
            <code className="text-slate-12">{baseEmailsDirectoryName}</code>{' '}
            folder.
          </Text>

          <Button asChild size="3">
            <Link href="https://react.email/docs">Check the docs</Link>
          </Button>
        </div>
      </div>
    </Shell>
  );
};

export default Home;
