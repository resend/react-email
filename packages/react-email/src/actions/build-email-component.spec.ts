import path from 'node:path';
import {
  buildEmailComponent,
  getEmailComponent,
} from '../actions/build-email-component';
import * as EnvExports from '../app/env';
import { isErr, isOk } from '../utils/result';

function mockEnv(emailsDirectoryAbsolutePath: string) {
  vi.resetModules();
  vi.spyOn(EnvExports, 'emailsDirectoryAbsolutePath', 'get').mockReturnValue(
    emailsDirectoryAbsolutePath,
  );
}

describe('getEmailComponent()', () => {
  describe('Node internals support', () => {
    const emailSlug = 'request-response-email';
    beforeAll(() => {
      mockEnv(path.resolve(__dirname, './testing'));
    });

    test('Request', async () => {
      const result = await buildEmailComponent(emailSlug);

      expect(isOk(result), 'There should be no errors building').toBe(true);
    });
  });

  test('with a demo email template', async () => {
    mockEnv(path.resolve(__dirname, '../../../../apps/demo/emails'));

    const emailSlug = 'notifications/vercel-invite-user';
    const result = await buildEmailComponent(emailSlug);

    if (isErr(result)) {
      console.log(result.error);
      expect('error' in result).toBe(false);
    } else {
      const emailComponent = await getEmailComponent(emailSlug);
      expect(emailComponent).toBeTruthy();
      expect(emailComponent!.sourceMap).toBeTruthy();

      const emailHtml = await emailComponent!.render(
        emailComponent!.createElement(
          emailComponent!.emailComponent,
          emailComponent!.emailComponent.PreviewProps,
        ),
      );
      expect(emailHtml).toMatchSnapshot();
    }
  });
});
