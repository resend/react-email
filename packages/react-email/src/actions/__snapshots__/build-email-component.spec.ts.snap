// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getEmailComponent() > with a demo email template 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html dir="ltr" lang="en"><head><link rel="preload" as="image" href="/static/vercel-logo.png"/><link rel="preload" as="image" href="/static/vercel-user.png"/><link rel="preload" as="image" href="/static/vercel-arrow.png"/><link rel="preload" as="image" href="/static/vercel-team.png"/><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"/><meta name="x-apple-disable-message-reformatting"/><!--$--></head><body style="background-color:rgb(255,255,255);margin-top:auto;margin-bottom:auto;margin-left:auto;margin-right:auto;font-family:ui-sans-serif, system-ui, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;, &quot;Noto Color Emoji&quot;;padding-left:0.5rem;padding-right:0.5rem"><div style="display:none;overflow:hidden;line-height:1px;opacity:0;max-height:0;max-width:0">Join Alan on Vercel<div> ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿ ‌​‍‎‏﻿</div></div><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="border-width:1px;border-style:solid;border-color:rgb(234,234,234);border-radius:0.25rem;margin-top:40px;margin-bottom:40px;margin-left:auto;margin-right:auto;padding:20px;max-width:465px"><tbody><tr style="width:100%"><td><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="margin-top:32px"><tbody><tr><td><img alt="Vercel" height="37" src="/static/vercel-logo.png" style="margin-top:0px;margin-bottom:0px;margin-left:auto;margin-right:auto;display:block;outline:none;border:none;text-decoration:none" width="40"/></td></tr></tbody></table><h1 style="color:rgb(0,0,0);font-size:24px;font-weight:400;text-align:center;padding:0px;margin-top:30px;margin-bottom:30px;margin-left:0px;margin-right:0px">Join <strong>Enigma</strong> on <strong>Vercel</strong></h1><p style="color:rgb(0,0,0);font-size:14px;line-height:24px;margin:16px 0">Hello <!-- -->alanturing<!-- -->,</p><p style="color:rgb(0,0,0);font-size:14px;line-height:24px;margin:16px 0"><strong>Alan</strong> (<a href="mailto:alan.turing@example.com" style="color:rgb(37,99,235);text-decoration-line:none" target="_blank">alan.turing@example.com</a>) has invited you to the <strong>Enigma</strong> team on<!-- --> <strong>Vercel</strong>.</p><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation"><tbody><tr><td><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation"><tbody style="width:100%"><tr style="width:100%"><td align="right" data-id="__react-email-column"><img height="64" src="/static/vercel-user.png" style="border-radius:9999px;display:block;outline:none;border:none;text-decoration:none" width="64"/></td><td align="center" data-id="__react-email-column"><img alt="invited you to" height="9" src="/static/vercel-arrow.png" style="display:block;outline:none;border:none;text-decoration:none" width="12"/></td><td align="left" data-id="__react-email-column"><img height="64" src="/static/vercel-team.png" style="border-radius:9999px;display:block;outline:none;border:none;text-decoration:none" width="64"/></td></tr></tbody></table></td></tr></tbody></table><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="text-align:center;margin-top:32px;margin-bottom:32px"><tbody><tr><td><a href="https://vercel.com/teams/invite/foo" style="background-color:rgb(0,0,0);border-radius:0.25rem;color:rgb(255,255,255);font-size:12px;font-weight:600;text-decoration-line:none;text-align:center;padding-left:1.25rem;padding-right:1.25rem;padding-top:0.75rem;padding-bottom:0.75rem;line-height:100%;text-decoration:none;display:inline-block;max-width:100%;mso-padding-alt:0px;padding:12px 20px 12px 20px" target="_blank"><span><!--[if mso]><i style="mso-font-width:500%;mso-text-raise:18" hidden>&#8202;&#8202;</i><![endif]--></span><span style="max-width:100%;display:inline-block;line-height:120%;mso-padding-alt:0px;mso-text-raise:9px">Join the team</span><span><!--[if mso]><i style="mso-font-width:500%" hidden>&#8202;&#8202;&#8203;</i><![endif]--></span></a></td></tr></tbody></table><p style="color:rgb(0,0,0);font-size:14px;line-height:24px;margin:16px 0">or copy and paste this URL into your browser:<!-- --> <a href="https://vercel.com/teams/invite/foo" style="color:rgb(37,99,235);text-decoration-line:none" target="_blank">https://vercel.com/teams/invite/foo</a></p><hr style="border-width:1px;border-style:solid;border-color:rgb(234,234,234);margin-top:26px;margin-bottom:26px;margin-left:0px;margin-right:0px;width:100%;border:none;border-top:1px solid #eaeaea"/><p style="color:rgb(102,102,102);font-size:12px;line-height:24px;margin:16px 0">This invitation was intended for<!-- --> <span style="color:rgb(0,0,0)">alanturing</span>. This invite was sent from <span style="color:rgb(0,0,0)">204.13.186.218</span> <!-- -->located in<!-- --> <span style="color:rgb(0,0,0)">São Paulo, Brazil</span>. If you were not expecting this invitation, you can ignore this email. If you are concerned about your account&#x27;s safety, please reply to this email to get in touch with us.</p></td></tr></tbody></table><!--/$--></body></html>"`;
